libs/core/thread_support/include/hpx/thread_support/spinlock.hpp

```diff
@@ -44,19 +45,17 @@ namespace hpx { namespace util { namespace detail {
 
         void lock() noexcept
         {
-            // Optimistically assume the lock is free on the first try
-            if (!m.load(std::memory_order_relaxed) &&
-                !m.exchange(true, std::memory_order_acquire))
-                return;
-
-            // Wait for lock to be released without generating cache misses
-            // Similar implementation to hpx::lcos::local::spinlock
             unsigned k = 0;
-            while (m.load(std::memory_order_relaxed) ||
-                m.exchange(true, std::memory_order_acquire))
+            do
             {
+                // Wait for lock to be released without generating cache misses
+                // Similar implementation to hpx::lcos::local::spinlock
+                if (try_lock())
+                    return;
+
                 yield_k(k++);
-            }
+
+            } while (true);
```

Comment:
What's wrong with simply writing:
```
do {
    yield_k(k++);
} while (!try_lock());
```

Reply:
OK. I just tried compiler explorer to check how compiler evaluates `!try_lock()` and it worked as well. I was thinking that it will lead to the evaluation of exchange on each iteration and thus avoided writing that way.